rust/otap-dataflow/crates/otap/src/otap_batch_processor.rs

```diff
@@ -0,0 +1,698 @@
+        };
+        config.send_batch_max_size = max;
+
+        if let Some(limit) = config.metadata_cardinality_limit {
+            if limit < MIN_METADATA_CARDINALITY_LIMIT {
+                config.metadata_cardinality_limit = Some(MIN_METADATA_CARDINALITY_LIMIT);
+            }
+        }
+        let user_config = Arc::new(NodeUserConfig::new_processor_config(
+            OTAP_BATCH_PROCESSOR_URN,
+        ));
+        let proc = OtapBatchProcessor {
+            config,
+            current: Vec::new(),
+        };
+        Ok(ProcessorWrapper::local(proc, node, user_config, proc_cfg))
+    }
+
+    /// Returns true if the buffered items reach or exceed the emission threshold.
+    ///
+    /// MVP: threshold is based on count (number of messages). In future, switch to real
+    /// row counts for OtapArrowRecords.
+    fn size_triggers_emission(&self) -> bool {
+        self.current.len() >= self.config.send_batch_size
+    }
+
+    /// Partition buffered items by signal type when possible and emit them downstream.
+    /// Items that cannot be converted to OtapArrowRecords are passed through as-is.
+    async fn flush_current(
+        &mut self,
+        effect: &mut local::EffectHandler<OtapPdata>,
+    ) -> Result<(), EngineError<OtapPdata>> {
+        if self.current.is_empty() {
+            return Ok(());
+        }
+        let mut drained: Vec<OtapPdata> = std::mem::take(&mut self.current);
+
+        // Separate items that can be converted to OtapArrowRecords from passthrough
+        let mut convertible: Vec<OtapArrowRecords> = Vec::new();
+        let mut passthrough: Vec<OtapPdata> = Vec::new();
+        for item in drained.drain(..) {
+            match OtapArrowRecords::try_from(item.clone()) {
+                Ok(rec) => convertible.push(rec),
+                Err(_) => passthrough.push(item),
+            }
+        }
+
+        // If nothing convertible, just emit passthrough items and return.
+        if convertible.is_empty() {
+            for item in passthrough {
+                effect.send_message(item).await?;
+            }
+            return Ok(());
+        }
+
+        // Build output batches using low-level batching (split + concatenate per type)
+        let max = NonZeroU64::new(self.config.send_batch_max_size as u64);
+        let output_batches = match low_make_output_batches(max, convertible.clone()) {
+            Ok(v) => v,
+            Err(e) => {
+                // Fall back to passthrough of convertible records on error
+                effect
+                    .info(&format!(
+                        "OTAP batch processor: low-level batching failed: {e}; falling back"
+                    ))
+                    .await;
+                convertible
+            }
+        };
+
+        // Emit converted (records) first, then passthrough
+        for records in output_batches {
+            let pdata: OtapPdata = records.into();
+            effect.send_message(pdata).await?;
+        }
```

Comment:
If there's some "remainder batch" after splitting/merging whose size is less than the max batch length, do we want to forward it immediately here, or re-buffer it so it can be combined with following batches?

I'm honestly not sure what the right answer is here because I guess that batch could have rows that need to be flushed due to some timer tick?